### PR TITLE
Show polls in timeline

### PR DESCRIPTION
**Pull request by julioromano**
- Creates `polls` module in `timeline` to hold all related logic.
  - `PollState` holds ruma types directly to keep it simple and reuse ruma’s existing aggregation logic.
  - `PollPendingEvents` acts as a cache for response/end events that are handled before their start event has been handled. 
- Handles ruma's `UnstablePoll{Start, Response, End}` events.
  - When a start event arrives it computes the starting `PollState`.
    - If there are responses/end events in the `PollPendingEvents` cache it will move them from the cache to the state.
  - When a response event arrives it updates the `PollState`.
    - Or it is added to the `PollPendingEvents` if no start event is found yet.
  - When an end event arrives it updates the `PollState`
    - Or it is added to the `PollPendingEvents` if no start event is found yet. 
- Handles retries.
- Handles edit events for the poll start event. 
- Adds timeline tests in `matrix_sdk_ui::timeline::tests::polls` (still to be improved).
- BREAKING: Removes the `matrix_sdk_ffi::timeline::TimelineItemContentKind::PollEnd` variant because the timeline won't show end events.

### CODE REVIEW

**Comment by poljar** [comment]
Does this PR replace https://github.com/matrix-org/matrix-rust-sdk/pull/2337?

**Comment by julioromano** [comment]
> Does this PR replace #2337?

Indeed. Closed that one.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:528]:
```suggestion
        update_timeline_item!(self, &c.relates_to.event_id.clone(), "poll response", |event_item| {
```

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:545]:
```suggestion
        update_timeline_item!(self, &c.relates_to.event_id.clone(), "poll end", |event_item| {
```

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:528]: (this is used for the debug message all the way at the end of this file)

File: crates/matrix-sdk-ui/src/timeline/mod.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:59]: Can you please make this module private like the other ones and re-export any public types from this module (as part of the `pub use` statement starting on line 71)?

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:148]: What is this impl used for? It feels a bit weird.

File: crates/matrix-sdk-ui/src/timeline/mod.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:59]: Sure I will, can you offer some background on the why to let me learn something new?

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:528]: I was actually wondering for what it was for but forgot to ask. Thanks.

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/polls.rs:148]: Yeah I agree, it's used only from test code, I'll move it in the test module

**Comment by julioromano** [comment]
Addressed all the comments except the first one (regarding caching response/end events related to missing start events).
That'll take more time.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:605]: Please actually copy-paste the body of `_update_timeline_item` instead of calling it here. It makes no sense for this method to log "Timeline item not found, discarding poll end" if the item is not found, since it won't actually be discarded, it will be added to the poll cache. We could put the found branch of `update_timeline_item` in a new function and reuse that though. Maybe *that* should be named `update_timeline_item`, and the other one `find_and_update_timeline_item`.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:560]:
```suggestion
            let new_content = match poll_state.edit(&replacement.new_content) {
                Ok(edited_poll_state) => TimelineItemContent::Poll(edited_poll_state),
                Err(e) => {
                    info!(
                        original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
                        "Failed to apply poll edit: {e}"
                    );
                    return None;
                }
            };
```

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:574]: Can you remove this TODO and instead make a new GitHub issue after this is merged, about being able to vote on your own poll while it is still in local-echo state? It would be nice to support eventually, but editing a local echo is likewise not yet supported, and it seems like a use case that's not terribly important.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:577]:
```suggestion
        }
```

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:4]:
```suggestion
//! This module handles rendering of MSC3381 polls in the timeline.

use imbl::HashMap;
```

**Review by zecakeh — COMMENTED** [review]
Could you try to change this to use https://github.com/ruma/ruma/pull/1627? It should simplify the code to compile results a lot.

Also I noted in the code that a lot of times functions are taking a `&MillisecondsSinceUnixEpoch` when they could take `MillisecondsSinceUnixEpoch`, since it implements `Copy`

**Comment by zecakeh** [comment]
Oh I also forgot, there should be some TODO/issue about handling live redacted responses and poll end events. There's no mention of it in the MSC, but, logically, since they lose their `m.relates_to` field, it should be as if it they were never received.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (zecakeh) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:608]: We should check first if the user has the right to end the poll. According to the MSC it is if its the creator of the poll, or they have the right to redact any event.

File: bindings/matrix-sdk-ffi/src/timeline.rs
Comment (julioromano) [reply bindings/matrix-sdk-ffi/src/timeline.rs:1317]:
```
error[E0308]: mismatched types
    --> bindings/matrix-sdk-ffi/src/timeline.rs:1319:20
     |
1319 |             votes: value.votes,
     |                    ^^^^^^^^^^^ expected `HashMap<String, Vec<String>>`, found a different `HashMap<String, Vec<String>>`
     |
     = note: `HashMap<String, Vec<String>>` and `HashMap<String, Vec<String>>` have similar names, but are actually distinct types
note: `HashMap<String, Vec<String>>` is defined in crate `imbl`
    --> /Users/marcor/.cargo/registry/src/index.crates.io-6f17d22bba15001f/imbl-2.0.0/./src/hash/map.rs:99:1
     |
99   | pub struct HashMap<K, V, S = RandomState> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `HashMap<String, Vec<String>>` is defined in crate `std`
    --> /Users/marcor/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:214:1
     |
214  | pub struct HashMap<K, V, S = RandomState> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:560]: Besides the different log message, why match better than if let in this context?

Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:560]: Besides, had to use `{e:?}`.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:574]: I don't think this is the case.
This is for the (perhaps even impossible) case where:
- responses / end events arrive before the start event, so they are cached.
- the start event eventually arrives but it is a local echo. Therefore we're not able to apply the cache to it.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:577]: what's the difference?

**Comment by julioromano** [comment]
Applied some of @jplatte suggestions, for a few I have further questions.

File: bindings/matrix-sdk-ffi/src/timeline.rs
Comment (zecakeh) [reply bindings/matrix-sdk-ffi/src/timeline.rs:1320]: Is there a reason you used `imbl::HashMap` here? I believe the one from `std` is used everywhere else.

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/polls.rs:85]: This would be the same as running the logic in `PollState.result()` every time `PollState.add_response()` or `PollState.end()` are called.

Therefore the answer depends on which methods are called more times on average usage:

`PollState.result()` (which is only called from in `matrix_sdk_ffi::timeline::TimelineItemContent.kind()` )

OR

`PollState.add_response()` and `PollState.end()` (which are called from `matrix_sdk_ui::timeline::event_handler::TimelineEventHandler.handle_event()`)

?

File: bindings/matrix-sdk-ffi/src/timeline.rs
Comment (julioromano) [reply bindings/matrix-sdk-ffi/src/timeline.rs:1320]: There you go :)
My noobness led to careless selection of ~~imports~~ `use` statements :)

**Comment by julioromano** [comment]
> > I noted in the code that a lot of times functions are taking a `&MillisecondsSinceUnixEpoch` when they could take `MillisecondsSinceUnixEpoch`, since it implements `Copy`
> 
> No problem in doing such a change, I just would like to know what's the rationale in doing this? Should I always blindly default to "copy" semantics for Copy types? Or when does it actually make sense to pass a reference to a Copy type?

I was actually also thinking if it was more idiomatic for those methods to accept `&UserId` instead of `&OwnedUserId` ?

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:572]: That is indeed not possible.

Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:572]: Good then, will remove the comment. I couldn't understand why this is not possible though :) Can you offer some background here?

**Comment by zecakeh** [comment]
> No problem in doing such a change, I just would like to know what's the rationale in doing this? Should I always blindly default to "copy" semantics for Copy types? Or when does it actually make sense to pass a reference to a Copy type?

I personally never pass a reference to a type that implements `Copy`. The idea behind a type that implements `Copy` is generally that it is inexpensive to duplicate, so not using a reference is just fine. However sometimes it is possible to get or need to pass a reference, when used with types/functions that work with generics.

If you want a probably better explanation, there is this clippy lint (that is not enabled by default): [trivially_copy_pass_by_ref](https://rust-lang.github.io/rust-clippy/stable/index.html#trivially_copy_pass_by_ref).

**Comment by jplatte** [comment]
> Should I always blindly default to "copy" semantics for Copy types? Or when does it actually make sense to pass a reference to a Copy type?

There are some cases where passing `Copy` types by reference is preferable (e.g. very large `Copy` types) or even necessary, but I'd say passing them by value should be the default. 

> I was actually also thinking if it was more idiomatic for those methods to accept `&UserId` instead of `&OwnedUserId` ?

Yes, didn't notice that `&OwnedUserId` was used.

**Comment by zecakeh** [comment]
> I was actually also thinking if it was more idiomatic for those methods to accept `&UserId` instead of `&OwnedUserId` ?

Yes, definitely :slightly_smiling_face:. I didn't do a full review, Jonas just asked me to look at the polls logic/API since I'm familiar with the MSCs.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:572]: I've actually added an `else` with an `info!()` just in case.

**Comment by julioromano** [comment]
> > Should I always blindly default to "copy" semantics for Copy types? Or when does it actually make sense to pass a reference to a Copy type?
> 
> There are some cases where passing `Copy` types by reference is preferable (e.g. very large `Copy` types) or even necessary, but I'd say passing them by value should be the default.
> 
> > I was actually also thinking if it was more idiomatic for those methods to accept `&UserId` instead of `&OwnedUserId` ?
> 
> Yes, didn't notice that `&OwnedUserId` was used.

Done! And what about in hashmaps? Should they contain `OwnedUserId` keys or can they contain `UserId` keys instead?

**Comment by jplatte** [comment]
> Done! And what about in hashmaps? Should they contain `OwnedUserId` keys or can they contain `UserId` keys instead?

`UserId` can't really be used on its own, it always has to be behind a reference or other indirection because it is a [dynamically-sized type](https://doc.rust-lang.org/reference/dynamically-sized-types.html).

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs]: None. I just find redundant semicolons after closing braces a bit confusing / weird.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs]: Local echoes are messages sent by the local user that have not been echoed back by the server yet. When a new one is added, that currently always means it was *just* created so there can't be any other events referencing it yet.

The only other case we will add in the future is restoring local echoes from disk, where I guess there will be a very slim chance that you closed the app before receiving the remote echo for the poll, but got some responses from the server before re-opening the timeline and loading it from disk. In that case, it doesn't really make sense to apply responses to the local echo since we know the full event (remote echo) exists further up in the timeline (so we could use this information to remove the local echo).

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:607]: > We should check first if the user has the right to end the poll. According to the MSC it is if its the creator of the poll, or they have the right to redact any event.

@jplatte on this regard, can you point me to some existing code that deals with checking power levels so I can "borrow" (rust pun intended) some wisdom from it?

**Comment by julioromano** [comment]
> Oh I also forgot, there should be some TODO/issue about handling live redacted responses and poll end events. There's no mention of it in the MSC, but, logically, since they lose their `m.relates_to` field, it should be as if it they were never received.

Such events should be handled at event_handler.rs:314 and be therefore ignored. Correct?
So what is the "TODO" here?

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:607]: https://github.com/matrix-org/matrix-rust-sdk/blob/4493cbf0ace49acf1667dbce6efb24ff8312551c/crates/matrix-sdk/src/room/mod.rs#L2025

**Comment by julioromano** [comment]
> > Such events should be handled at event_handler.rs:314 and be therefore ignored. Correct? So what is the "TODO" here?
> 
> If they are redacted _after_ being received, the `PollState` needs to be updated by removing those events from the state.

Oh it seems the two versions of the MSC conflict in this case.

From v1:
> Spoiled votes are also how users can "un-vote" from a poll - redacting, or setting answers to an empty array, will spoil that user's vote.

From v2:
> Only a user's most recent vote (by origin_server_ts) is accepted, even if that event is invalid or redacted. Votes with timestamps after the poll has closed are ignored, as if they never happened. Note that redaction currently removes the m.relates_to information from the event, causing the vote to be detached from the poll. In this scenario, the user's vote is reverted to its previous state rather than explicitly spoiled. To "unvote" or otherwise override the previous vote state, clients should send a response with an empty m.selections array.

IIUC:
- v1 allows users to "unvote" by either sending a redaction event of their previous vote OR by sending a response with empty selections array.
- v2 completely ignores redactions and says users that want to "unvote" should send a response with empty selections array.

**Comment by zecakeh** [comment]
> IIUC:
> 
>     * v1 allows users to "unvote" by either sending a redaction event of their previous vote OR by sending a response with empty selections array.
> 
>     * v2 completely ignores redactions and says users that want to "unvote" should send a response with empty selections array.

The v1 has actually an error that was fixed in v2. hen the even is redacted, it loses its `m.relates_to` so it's not possible to know which poll it belongs to. So it should just be ignored.

**Comment by julioromano** [comment]
> Could you try to change this to use https://github.com/ruma/ruma/pull/1627? It should simplify the code to compile results a lot. Also don't hesitate to comment on the PR if you think further changes would simplify your work.

Done in https://github.com/matrix-org/matrix-rust-sdk/pull/2381/commits/73e23ed64b6d03f6ee724ff31c3689526d1f98d4

File: Cargo.toml
Comment (zecakeh) [inline Cargo.toml:37]: Could you update to this rev: `eeef5555238c6ee38bee5711930b256d0255e627`?

It's the lax-id-validation branch we use for Ruma.

Comment (julioromano) [reply Cargo.toml:37]: Yes thanks, was waiting for it :)

File: bindings/matrix-sdk-ffi/src/timeline.rs
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/timeline.rs:1317]: Since this function takes ownership of the `PollResult`, we can take ownership of the answers here via `.into_iter()`, to avoid having to clone:

```suggestion
                .into_iter()
                .map(|i| PollAnswer { id: i.id, text: i.text })
```

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:550]: Indented too much.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:579]: There can't be pending events for this poll if the flow is `Local`, because new local echoes don't have an event ID that could be referenced by responses yet. I think it makes most sense to just remove the `else` branch.

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:149]:
```suggestion
        self.pending_poll_responses.entry(start_id.to_owned()).or_default().push(
```

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:166]:
```suggestion
            poll_state.response_data.append(pending_responses);
```

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:162]:
```suggestion
    /// Dumps all response and end events present in the cache that belong to
```

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:169]:
```suggestion
            poll_state.end_event_timestamp = Some(pending_end.clone());
```

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/polls.rs:73]:
```suggestion
    /// Marks the poll as ended.
    ///
    /// If the poll has already ended, returns `Err(())`.
```

File: crates/matrix-sdk-ui/src/timeline/tests/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/polls.rs:213]: Not sure if you've seen the clippy warnings about this from CI?

```suggestion
    assert_eq!(a.answers[0].id, b.answers[0].id);
    assert_eq!(a.answers[0].text, b.answers[0].text);
    assert_eq!(a.answers[1].id, b.answers[1].id);
    assert_eq!(a.answers[1].text, b.answers[1].text);
```

Although why not just loop over all the answers?

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:550]: Fixed. Not sure why most of the times vscode applies formatting automatically but sometimes not.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs:579]: Removed else and left a comment with your great explanation so I won't forget.

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (julioromano) [reply crates/matrix-sdk-ui/src/timeline/polls.rs:149]: Is there any difference besides style?

**Comment by julioromano** [comment]
There is a complication I didn't foresee with applying poll response redactions locally:
Whenever we get a `RoomRedactionEvent` we get (in the `redacts` field) the id of the event that needs to be locally redacted. Let's say it's the ID of a poll response.

To apply the redaction locally we must then remove the affected response either from the `PollPendingEvents` cache or from the `PollState` instance inside the `TimelineItemContent::Poll` item that response belongs to.

Problems:
1) There is no way to know upfront if a response is in the `PollPendingEvents` or in a `PollState` so we always have to search in both places.

2) There is no way to know upfront in which `PollState` instance a response belongs to. Only way is to iterate through all `TimelineItemContent::Poll`s and, for each, loop through all of its responses.

3) There is no way to know upfront whether a `redacts` field points to a poll response.

This means that for every `RoomRedactionEvent` we receive (no matter if it's redacting a poll event or a completely different event) we'll have to loop through all the responses in `PollPendingEvents` and through all the responses in all the `TimelineItemContent::Poll`s to find if there's a matching response to remove.
So we'll potentially do this seemingly complicated and expensive search for any incoming redaction.
Am I getting this right?


BTW: Do we also have to handle redactions of poll end events? :D

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (zecakeh) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs]: In this case that's probably because you don't have a recent enough version of rustfmt on your system. Iirc, support for formatting let-else statements has been added in one of the very latest versions.

**Comment by zecakeh** [comment]
> So we'll potentially do this seemingly complicated and expensive search for any incoming redaction.

You are right, I believe that's why we don't handle redaction of replied-to events currently (#2116). This could be fixed by having a map of relations we encounter for example, but I would leave this to a separate PR.

**Comment by julioromano** [comment]
Then after this last commit I guess this is pretty much done, thanks a lot for all the comments!
Let me know if there's any other improvements I can do.

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs]: Being able to inspect the error is the only thing that really makes a difference, apart from formatting / indentation. However, I had missed that this is not using a proper error type (just `()`), so the if-let actually makes sense, there is no error message to log here.

File: crates/matrix-sdk-ui/src/timeline/polls.rs
Comment (jplatte) [reply crates/matrix-sdk-ui/src/timeline/polls.rs]: There is a tiny difference in that before, the vector was instantiated regardless of whether the entry was found or not. However, since creating an empty vector doesn't allocate, it should not make an actual difference in runtime code, at least for release builds.

File: crates/matrix-sdk-ui/src/timeline/tests/polls.rs
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/polls.rs:183]:
```suggestion
    async fn event_items(&self) -> Vec<EventTimelineItem> {
```

**Review by jplatte — APPROVED** [review]

**Comment by julioromano** [comment]
@jplatte @zecakeh 🥳
This was my first Rust contribution ever! Thank you very much for all the help provided! I enjoyed this learning journey and look forward to more.

**Pull request merged**
